### PR TITLE
Jonmv/async document v1

### PR DESCRIPTION
**Pull request by jonmv**
@bjorncs or @vekterli please review.

### CODE REVIEW

File: documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java [hunk 1]
```diff
@@ -55,8 +55,11 @@
      * while attempting to create such an object, this method will throw an
      * exception.
      *
+     * @deprecated Inject a DocumentManagerConfig and create a MessageBusDocumentAccess from this instead.
```
Comment (bratseth) [inline documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: But that's more difficult ... doesn't this just work?

**Review by bratseth — APPROVED** [review]
LGTM apart from that question

File: documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java (hunk 1 above)
Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: @baldersheim wqs of the opinion this should go.

Comment (baldersheim) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: It uses self subscription.
1 - It means that at will change out of sync with the other components also using the same config.
   - Components will be out of sync in for a short time.
   - It is either very costly due to extra synchronisation. However I think it is worse, they will never agree, as there are thread visibility issues.
2 - It leaks resources. Every heapdump I have checked where the customer has done this, it is leaking threads and other resources.

3 - I think that having the config injected is a simple remediation for this mess. I thought that was the whole reason for having config injected and creating final components.

This is one og our bad past patterns.

Comment (bratseth) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: Yes DocumentAccess is a factory for clients, if you're in a container you should have everything injected. This used to be convenient when running a client on a node which can reach config servers. It's not relevant to cloud deployments, so fine I guess.

Btw, we should have an object you can have injected directly to make it more obvious what to do when running in a container, perhaps that could be done as part of this work.

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: Yes, I thought the same thing. The document access is perhaps something we wish to create only when needed, so a factory which is always present, based on the relevant config?

Comment (bratseth) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: Yes, or preferably something like a session (or, two - one sync and one async) which is always present so we don't need to destroy it - but otoh I'm not sure if we want to create the connections if we don't need them.

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: But sync sessions are often used one per client? So rather a session factory, i.e. a document access? How about a lazy doc access, then? If one of those is always enough.

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: A doc access is also what you need for visitor sessions.

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: Hmm, why the sync sessions are used one per thread/operation in the current document/v1, I don't know ... Seems like an error to me. So you could indeed have injectable `Sync` and `Async`-`Session`s. That doesn't cover the other types of sessions, though.

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: If anything, using multiple `SyncSessions` to handle a feed of operations breaks operation sequencing ...

Comment (bratseth) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: Yes, that should work if we're ok with always having instances of these sessions. 

What other types of sessions?

Comment (jonmv) [reply documentapi/src/main/java/com/yahoo/documentapi/DocumentAccess.java:58]: VisitorSession. I guess SubscriptionSession never got an implementation. See #14461 for a suggestion of a lazy `DocumentAccess` provider.

File: documentapi/src/main/java/com/yahoo/documentapi/DocumentIdResponse.java [hunk 2]
```diff
@@ -36,9 +35,9 @@ public DocumentIdResponse(long requestId, DocumentId documentId) {
      * @param textMessage the message to encapsulate in the Response
      * @param success     true if the response represents a successful call
      */
+    @Deprecated(since = "7") // TODO: Remove on Vespa 8
```
Comment (bjorncs) [inline documentapi/src/main/java/com/yahoo/documentapi/DocumentIdResponse.java:38]: Consider setting `forRemoval=true`. IDE/compiler gives stronger feedback if true.

**Review by bjorncs — COMMENTED** [review]
👍

**Pull request merged**